src/runge_kutta.jl

```diff
@@ -0,0 +1,509 @@
+        for s=1:S
+            for d=1:dof # loop as ytmp[:] = ys[:,i] allocates!
+                ytmp[d] = ys[d,i]
+            end
+            calc_next_k!(ks, ytmp, ytmp, s, fn, tspan[i], dt, dof, btab)
+            for d=1:dof
+                ys[d,i+1] += dt * btab.b[s]*ks[d,s]
+            end
+        end
+    end
+    return tspan, ys
+end
+
+# calculates k[s]
+function calc_next_k!{N,S}(ks::Matrix, ytmp::Vector, y, s, fn, t, dt, dof, btab::TableauRKExplicit{N,S})
+    # Calculates the next ks and puts it into ks[:,s]
+    # - ks and ytmp are modified inside this function.
+    ytmp[:] = y
+    for ss=1:s-1, d=1:dof
+        ytmp[d] += dt * ks[d,ss] * btab.a[s,ss]
+    end
+    ks[:,s] = fn(t + btab.c[s]*dt, ytmp)
+    nothing
+end
+
+ode4_v2(fn, y0, tspan) = oderk_fixed(fn, y0, tspan, bt_rk4)
+ode1_euler(fn, y0, tspan) = oderk_fixed(fn, y0, tspan, bt_feuler)
+ode2_midpoint(fn, y0, tspan) = oderk_fixed(fn, y0, tspan, bt_midpoint)
+ode2_heun(fn, y0, tspan) = oderk_fixed(fn, y0, tspan, bt_heun)
+
+
+# Adaptive ODE time stepper
+function oderk_adapt(fn, y0, tspan, btab::TableauRKExplicit; kwords...)
+    # For y0 which don't support indexing.
+    fn_ = (t, y) -> fn(t, y[1])
+    t,y = oderk_adapt(fn_, [y0], tspan, btab; kwords...)
+    return t, y[:]
+end
+function oderk_adapt{N,S}(fn, y0::AbstractVector, tspan, btab_::TableauRKExplicit{N,S};
+                          reltol = 1.0e-5, abstol = 1.0e-8,
+#                          norm=Base.norm,
+                          minstep=abs(tspan[end] - tspan[1])/1e9,
+                          maxstep=abs(tspan[end] - tspan[1])/2.5,
+                          initstep=0.,
+                          points=:all
+                          )
+    # For y0 which support indexing.  Currently y0<:AbstractVector but
+    # that could be relaxed with a Holy-trait.
+    !isadaptive(btab_) && error("Can only use this solver with an adaptive RK Butcher table")
+
+    Et, Ey, Ef, btab = make_consistent_types(fn, y0, tspan, btab_)
+    # parameters
+    order = minimum(btab.order)
+    timeout_const = 5 # after step reduction do not increase step for
+                      # timeout_const steps
+
+    ## Initialization
+    dof = length(y0)
+    tspan = convert(Vector{Et}, tspan)
+    t = tspan[1]
+    tstart = tspan[1]
+    tend = tspan[end]
+
+    # work arrays:
+    y      = similar(y0, Ey, dof)      # y at time t
+    y[:]   = y0
+    ytrial = similar(y0, Ey, dof) # trial solution at time t+dt
+    yerr   = similar(y0, Ey, dof) # error of trial solution
+    ks     = similar(y0, Ey, dof, S)
+    ytmp   = similar(y0, Ey, dof)
+    f0     = similar(y0, Ey, dof)
+    f1     = similar(y0, Ey, dof)
+
+    # Option points determines where solution is returned:
+    if points==:specified
+        nsteps = length(tspan)
+        ys = similar(y0, Ey, dof, nsteps)
+        ys[:,1] = y
+        tspan_fixed = []
+    elseif points==:all
+        ys = similar(y0, Ey, 0)
```

Comment:
Actually `sizehint` will only allocate the memory needed for the pointers to the vectors, not the space needed for the vectors themselves.